Allow specifying styles as `Map`s to guarantee ordering

xymostech opened the pull request:
Summary: Key ordering in objects can be different in different environments.
Sometimes, this causes problems, like where styles generated on the server are
not in the same order as the same styles generated on the client. This
manifests in problems such as #199

This change lets users manually fix instances where the ordering of elements
changes by specifying their styles in an ES6 `Map`, which has defined value
ordering.

In order to accomplish this, an `OrderedElements` class was created, which is
sorta like a `Map` but can only store string keys and lacks most of the
features. Internally, `Map`s and objects are converted into this and then these
are merged together to preserve the ordering.

Fixes #199

Test Plan:
 - `npm test`

@lencioni @ljharb

xymostech commented:
I haven't written docs/examples for this yet, just wanted to throw this out there to see if this sounds like a reasonable solution to the problem you're seeing in #199.

I don't think your code in particular would have to change in order for this to make your issue work correctly, since you're already implicitly specifying the ordering you want via two different style objects. If someone was having problems with e.g.

```
StyleSheet.create({
    a: {
        margin: 0,
        marginLeft: 15,
    },
})
```

turning into

```
margin-left: 15px;
margin: 0px;
```

then they could specify their styles via a `Map` like

```
StyleSheet.create({
    a: new Map([
        ["margin", 0],
        ["marginLeft", 15],
    ]),
})
```

Maybe that's not actually something we should worry about, but I don't know.

lencioni commented on package.json:
```diff
@@ -44,6 +44,7 @@
     "babel-core": "^5.8.25",
     "babel-loader": "^5.3.2",
     "chai": "^3.3.0",
+    "core-js": "^2.4.1",
```
Are you bringing this in for Map? It might be a little bit nicer to bring in something a little more scoped like https://github.com/airbnb/browser-shims or just https://www.npmjs.com/package/es6-shim

ljharb replied:
`core-js`'s `Map` is not spec-compliant, since it mutates object keys in order to achieve otherwise-impossible performance. I highly recommend using the es6-shim instead here.

xymostech replied:
Interesting! This is specifically for the few tests which are using `Map`, not for the actual library, so I'm not super worried about spec compliance here. I can switch it out, though.

lencioni commented on src/generate.js:
```diff
@@ -143,18 +144,20 @@ export const generateCSS = (
     let generatedStyles = "";
 
-    Object.keys(merged).forEach(key => {
+    merged.forEach((key, val) => {
```
unrelated: Although I generally think stuff like `forEach` is best, regular for loops are faster which might matter in performance-sensitive applications. It might be worth looking into optimizing Aphrodite in this way at some point.

ljharb replied:
(please do so with benchmarks, however, and not just automatically)

xymostech replied:
Yeah, that's definitely a good point. I'll put in a TODO to think about this later. We haven't done any serious benchmarking yet, so right now I think there are other places that could benefit more from optimization.

lencioni commented on src/generate.js:
```diff
@@ -244,44 +242,36 @@ const runStringHandlers = (
+    // of the rules aren't in the original list, we sort them to the top.
+    prefixedRules.sort((a, b) => {
+        const aIndex = handledDeclarations.keyOrder.indexOf(a);
+        const bIndex = handledDeclarations.keyOrder.indexOf(b);
```
It would be nice to use something with constant time lookup here to avoid all of the array scans in this loop. A simple object would probably do the trick.

lencioni commented on src/ordered-elements.js:
```diff
@@ -0,0 +1,77 @@
+    }
+}
+
+OrderedElements.fromObject = (obj) => {
```
I wonder if this should take an optional array argument for `keyOrder` so it can be specified explicitly when possible.

xymostech replied:
At that point, you'd basically just be calling the constructor, so I'm not sure I see the usefulness of this. Anyway, we're not using that for now; we can add it if it's necessary.

lencioni commented:
Also, thanks for the fast response on this! 🎆

xymostech commented:
Thanks for the review, @lencioni! I'll look into making some of the changes you mentioned.

Not sure if this is possible for you, but would it be feasible for you to pull this change into your codebase to make sure that it actually fixes your problem? I'd hate to ship this only to see the bug still manifesting. :P

lencioni commented:
Yes, but I almost certainly won't get to it until Monday at the earliest

lencioni commented:
I've tested this out in my dev environment and it seems to resolve the issue we were running into!

xymostech commented on src/generate.js:
```diff
@@ -186,31 +190,25 @@ export const generateCSS = (
     stringHandlers /* : StringHandlers */,
     selectorHandlers /* : SelectorHandler[] */
 ) /* */ => {
-    const result = {};
```
:( I undid some of your for-loop optimization here, but putting it back is lot uglier now with the `OrderedElements` thing. I re-did some of the optimizations inside of the `OrderedElements` implementation, though.

lencioni replied:
No worries, we can always come back, profile, and optimize the hotspots again. 🚀

xymostech commented:
@lencioni @ljharb I think I addressed all of the things you mentioned, and I added some docs now too. Mind taking one more look at it before I land? I think once this and the rest of your diffs are landed @lencioni I'll release a new version.

lencioni commented on README.md:
````diff
@@ -429,6 +429,80 @@ then we end up with the opposite effect, with `foo` overriding `bar`! The way to
+   });
+   ```
+
+   Note that `Map`s are not fully supported in all browsers.
````
It might be worth linking off to some resources here.

- https://github.com/airbnb/browser-shims
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map

xymostech replied:
the mdn link is up above, I'll include a link to the shim here.

lencioni commented on src/generate.js:
```diff
@@ -186,31 +190,25 @@ export const generateCSS = (
-            result[keys[i]] = stringHandlers[keys[i]](
-                declarations[keys[i]], selectorHandlers);
+            return stringHandlers[key](val, selectorHandlers);
         } else {
```
nit: you are returning early on 207, so you can remove the `else` here and move `return val;` up to the top level.

lencioni commented on src/generate.js:
```diff
@@ -246,44 +244,43 @@ const runStringHandlers = (
+    const sortOrder = {};
+    for (let i = 0; i < prefixedRules.length; i++) {
+        const key = prefixedRules[i][0];
+        sortOrder[key] = handledDeclarations.keyOrder.indexOf(key);
```
It would be nice to avoid the array scan in this loop too.

lencioni commented on src/generate.js:
```diff
@@ -246,44 +244,43 @@ const runStringHandlers = (
+    // and we sort them to the top.
+    prefixedRules.sort((a, b) => {
+        return sortOrder[a[0]] - sortOrder[b[0]];
+    });
```
I'm a little worried about this possibly changing the order of some prefixed rules in a problematic way. e.g. if you have a shorthand style and a longhand style of the same property that both need to be prefixed, will the prefixes preserve order? We might need to handle this a bit more delicately.

xymostech replied:
That's a good point. I'm not sure how to do this perfectly without changing the format that inline-style-prefixer outputs...

What if we do something like sort all of `key`, `"-webkit-" + key`, `"-moz-" + key`, and `"-ms-" + key` to the same position in the list? There are probably some prefixes we will miss which we'll put at the top, but that should catch most things?

lencioni replied:
I think that might be okay as long as we only do this for prefixed styles that do not have a sortOrder. In other words, if I explicitly pass in a prefixed style after its unprefixed style, that order should be maintained.

xymostech replied:
Good point! Thanks for catching that. Should be fixed in the newest version.

lencioni commented on src/ordered-elements.js:
```diff
@@ -0,0 +1,82 @@
+    set(key /* : string */, value /* : any */) {
+        if (!this.elements.hasOwnProperty(key)) {
+            this.keyOrder.push(key);
+        }
```
If the element is already there, should it move it to the end in keyOrder?

xymostech replied:
I don't think so. Or at least, that would be semantically different than what we've had before. That is, merging

```
{
    a: 1,
    b: 2,
}
```

with

```
{
    a: 3,
}
```

gives

```
{
    a: 3,
    b: 2,
}
```

now. Moving to the end of the `keyOrder` would change the merged result to

```
{
    b: 2,
    a: 3,
}
```

lencioni replied:
That's right. I'm wondering if the before behavior is unexpected and could cause similar bugs?

xymostech replied:
In the current way, it's impossible to make new styles appear after the old styles. Though if we switch it, it'll be impossible to make new styles appear before the old styles. I think either way has problems, but this at least stays consistent with what we have now.

lencioni commented on src/ordered-elements.js:
```diff
@@ -0,0 +1,82 @@
+
+OrderedElements.from = (obj) => {
+    if (obj instanceof OrderedElements) {
+        return obj;
```
Should this create a new instance? That would be slower, but probably more correct.
